components/ide/jetbrains/backend-plugin/src/main/kotlin/io/gitpod/jetbrains/remote/GitpodTerminalService.kt

```diff
@@ -0,0 +1,162 @@
+            val terminalsList = getSupervisorTerminalsListAsync().await().terminalsList
+            debug("Got a list of Supervisor terminals: $terminalsList")
+            runInEdt {
+                if (tasksList.isEmpty()) {
+                    backendTerminalManager.createNewSharedTerminal("Gitpod", "Terminal")
+                } else {
+                    val aliasToTerminalMap: MutableMap<String, TerminalOuterClass.Terminal> =
+                            mutableMapOf()
+
+                    for (terminal in terminalsList) {
+                        val terminalAlias = terminal.alias
+                        aliasToTerminalMap[terminalAlias] = terminal
+                    }
+
+                    for (task in tasksList) {
+                        val terminalAlias = task.terminal
+                        val terminal = aliasToTerminalMap[terminalAlias]
+
+                        if (terminal != null) {
+                            createSharedTerminal(terminal)
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    private fun getTerminalToolWindowRegisteredEvent(): CompletableFuture<Void> {
+        val completableFuture = CompletableFuture<Void>()
+
+        debug("Waiting for TerminalToolWindow to be registered...")
+        val toolWindowManagerListener =
+                object : ToolWindowManagerListener {
+                    override fun toolWindowsRegistered(
+                            ids: MutableList<String>,
+                            toolWindowManager: ToolWindowManager
+                    ) {
+                        if (ids.contains(TerminalToolWindowFactory.TOOL_WINDOW_ID)) {
+                            debug("TerminalToolWindow got registered!")
+                            completableFuture.complete(null)
+                        }
+                    }
+                }
+
+        project.messageBus
+                .connect()
+                .subscribe(ToolWindowManagerListener.TOPIC, toolWindowManagerListener)
+
+        return completableFuture
+    }
+
+    private fun getSupervisorTasksList(): CompletableFuture<List<TaskStatus>> {
+        val completableFuture = CompletableFuture<List<TaskStatus>>()
+        val taskStatusRequest = TasksStatusRequest.newBuilder().setObserve(true).build()
+        val taskStatusResponseObserver =
+                object : StreamObserver<TasksStatusResponse> {
+                    override fun onNext(response: TasksStatusResponse) {
+                        debug("Received task list: ${response.tasksList}")
+
+                        var hasOpenedAllTasks = true
+
+                        response.tasksList.forEach { task ->
+                            if (task.state === TaskState.opening) {
+                                hasOpenedAllTasks = false
+                            }
+                        }
+
+                        if (hasOpenedAllTasks) {
+                            this.onCompleted()
```

Comment:
are you sure that it is enough to stop streaming? I think you need to implement `beforeStart` to get an acess to `ClientCallStreamObserver<SubscribeRequest>` and be able to call cancel on it

Reply:
Yes, according to the [docs](https://grpc.io/docs/languages/java/basics/):
"We use the response observer’s `onCompleted()` method to specify that we’ve finished dealing with the RPC."

But in any case, I tried calling `clientCallStreamObserver.cancel("Reason Here", null)` but it didn't work well, as when we cancel through this method, it invokes `onError`.